### ISSUE TITLE
Return type inspection should not suggest self for @return $this

### ISSUE THREAD

**Issue by cautionbug** [kalessil/phpinspectionsea#1267]
i reported this over at [PhpStorm](https://youtrack.jetbrains.com/issue/WI-46044), but they corrected me that it's an EA inspection issue...

#761 seems to be a semantically incorrect change.

The context of `$this` is the last-extending class, whereas `self` is the current class. Applying `: self` incorrectly breaks fluency and extensibility of functions because they become forced to return an ancestor rather than the current instance. Consider the following:

```php
declare(strict_types = 1);        

class A {
    /**
     * Correct fluency
     * @return $this
     */
    public function foo() { return $this; }
}
class B extends A {
    public function foo() {
        // Do some extended stuff.

        // Return the current instance $this
        return parent::foo();
    }
}

class C {
    /**
     * Broken fluency
     * @return $this
     */
    public function foo(): self { return $this; }
}
class D extends C {
    // Signature must match, forcing return of B.
    public function foo(): self {
        // Do some extended stuff.

        // Return the current instance $this
        return parent::foo();
    }
}

(new B)->foo(); // Clean
(new D)->foo(); // Fatal
```

When the PhpDoc annotation indicates `@return $this`, a return type should not be suggested: `self` is the wrong context, and `static`/`$this` are invalid. The `: self` return suggestion should be offered only when the annotation indicates `@return self` or `@return CurrentClassName`.

**Comment by kalessil** [comment]
Please check the inspection settings: unchecking `Respect `@return self`` should do the trick.

**Comment by cautionbug** [comment]
Disabling an inspection doesn't make the inspection correct... i'm just avoiding the inspection's fault.

**Comment by cautionbug** [comment]
i've found the inspection option. i guess i'm unclear why that option changes behavior regarding `@return $this`

**Comment by kalessil** [comment]
Misleading wording, typical OSS/software issue =) I'll update the inspection title.

**Comment by kalessil** [comment]
You re welcome =) When wording is badly done, I'm happy to improve it =)
